subsys/bluetooth/host/att.c

```diff
@@ -2837,26 +2898,40 @@ static int bt_att_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
 		}
 	}
 
+	/* Thread-local variable, shouldn't be used by anything else */
+	__ASSERT_NO_MSG(!att_chan->rsp_buf);
+
+	/* Mark buffer free for re-use by the opcode handler.
+	 *
+	 * This allows ATT to always be able to send a RSP (or err RSP)
+	 * to the peer, regardless of the TX buffer usage by other stack
+	 * users (e.g. GATT notifications, L2CAP using global pool, SMP,
+	 * etc..), avoiding an ATT timeout due to resource usage.
+	 *
+	 * The ref is taken by `bt_att_chan_create_pdu`.
```

Comment:
It does take a ref:
https://github.com/jori-nordic/zephyr/blob/a391f85f3deec1370fe98ca1272e1169b0a03fc4/subsys/bluetooth/host/att.c#L720

This `rsp_buf` is only ever read/used by `bt_att_chan_create_pdu`. And when it is, we take a reference (see above). 

The ref take used to be where you want it but that was problematic: 
Some RSP handlers will try to allocate another buffer for the err rsp if the rsp handling fails. 
Doing so, it first does a first `unref`, so the ref we'd take here is consumed, then it 'allocates' another one, but whatever is in `rsp_buf` ends up being used again. Then before returning from `bt_att_recv`, we'd `unref`, thinking we just unrefed the ref we took at the start -> buffer is destroyed and lower layers freak out.

It's a bit hard to explain, but I can try to reproduce the failure and attach the logs/stacktrace if you're unconvinced.

What I did here is take the reference where the buffer is "produced": if `bt_att_chan_create_pdu` is the one creating the buffer from ATT's point of view. So I don't have to change the semantics of all the att buffer handling. They can allocate and unref at will and it will just work.

This `att_chan->rsp_buf = buf;` line only stores the "option" to steal the REQ buffer. If it ends up stolen, then whatever steals it `bt_att_chan_create_pdu` in this case will take the ref.

Reply:
that said, I can probably take another ref just in case, that shouldn't be a problem

Reply:
seems like I remembered wrong, the tests pass with the extra ref. Lemme know if that's good enough